lib/spack/spack/store.py

```diff
@@ -84,3 +85,28 @@ def _store():
 root = llnl.util.lang.LazyReference(lambda: store.root)
 db = llnl.util.lang.LazyReference(lambda: store.db)
 layout = llnl.util.lang.LazyReference(lambda: store.layout)
+
+
+def retrieve_upstream_dbs():
+    other_spack_instances = spack.config.get(
+        'config:upstream_spack_installations') or []
+
+    install_roots = []
+    for install_properties in other_spack_instances:
+        install_roots.append(install_properties['spack-install-prefix'])
+
+    return _construct_upstream_dbs_from_install_roots(install_roots)
+
+
+def _construct_upstream_dbs_from_install_roots(
+        install_roots, _test=False):
```

Comment:
I don’t like that we have separate code paths for testing… I’d rather somehow capture the warning output and fail the test based on that, but separate code paths scare me from a maintenance perspective and an accuracy of testing perspective.

Reply:
I may be missing something: this function is tested and also used by `retrieve_upstream_dbs`, so I don't regard the code paths as separate. This function isolates the most-easily-tested logic. If you are referring to the `_test` variable, the only effect it has right now is to throw an error in place of printing a warning, so that the testing framework can confirm the behavior. Was that still problematic or was there another concern I missed?

Reply:
This still concerns me, but it doesn't need to change as part of this PR.